test/e2e/node/pod_resize.go

```diff
@@ -1264,31 +1309,40 @@ func doPodResizeTests() {
 				}
 			}
 
-			ginkgo.By("patching pod for resize")
-			patchedPod, pErr = f.ClientSet.CoreV1().Pods(newPod.Namespace).Patch(context.TODO(), newPod.Name,
-				types.StrategicMergePatchType, []byte(tc.patchString), metav1.PatchOptions{})
-			framework.ExpectNoError(pErr, "failed to patch pod for resize")
-
-			ginkgo.By("verifying pod patched for resize")
-			verifyPodResources(patchedPod, tc.expected)
-			verifyPodAllocations(patchedPod, tc.containers, true)
+			patchAndVerify := func(patchString string, expectedContainers []TestContainerInfo, initialContainers []TestContainerInfo, opStr string) {
+				ginkgo.By(fmt.Sprintf("patching pod for %s", opStr))
+				patchedPod, pErr = f.ClientSet.CoreV1().Pods(newPod.Namespace).Patch(context.TODO(), newPod.Name,
+					types.StrategicMergePatchType, []byte(patchString), metav1.PatchOptions{})
+				framework.ExpectNoError(pErr, fmt.Sprintf("failed to patch pod for %s", opStr))
+
+				ginkgo.By(fmt.Sprintf("verifying pod patched for %s", opStr))
+				verifyPodResources(patchedPod, expectedContainers)
+				verifyPodAllocations(patchedPod, initialContainers, true)
+
+				ginkgo.By(fmt.Sprintf("waiting for %s to be actuated", opStr))
+				resizedPod := waitForPodResizeActuation(f.ClientSet, podClient, newPod, patchedPod, expectedContainers)
+
+				// Check cgroup values only for containerd versions before 1.6.9
+				if !isInPlaceResizeSupportedByRuntime(f.ClientSet, newPod.Spec.NodeName) {
+					ginkgo.By(fmt.Sprintf("verifying pod container's cgroup values after %s", opStr))
+					if !framework.NodeOSDistroIs("windows") {
+						verifyPodContainersCgroupValues(resizedPod, expectedContainers, true)
+					}
+				}
 
-			ginkgo.By("waiting for resize to be actuated")
-			resizedPod := waitForPodResizeActuation(f.ClientSet, podClient, newPod, patchedPod, tc.expected)
+				ginkgo.By(fmt.Sprintf("verifying pod resources after %s", opStr))
+				verifyPodResources(resizedPod, expectedContainers)
 
-			// Check cgroup values only for containerd versions before 1.6.9
-			if !isInPlaceResizeSupportedByRuntime(f.ClientSet, newPod.Spec.NodeName) {
-				ginkgo.By("verifying pod container's cgroup values after resize")
-				if !framework.NodeOSDistroIs("windows") {
-					verifyPodContainersCgroupValues(resizedPod, tc.expected, true)
-				}
+				ginkgo.By(fmt.Sprintf("verifying pod allocations after %s", opStr))
+				verifyPodAllocations(resizedPod, expectedContainers, true)
 			}
 
-			ginkgo.By("verifying pod resources after resize")
-			verifyPodResources(resizedPod, tc.expected)
+			patchAndVerify(tc.patchString, tc.expected, tc.containers, "resize")
 
-			ginkgo.By("verifying pod allocations after resize")
-			verifyPodAllocations(resizedPod, tc.expected, true)
+			rbPatchStr, err := genPatchString(tc.containers)
+			framework.ExpectNoError(err)
+			// Resize has been actuated, test rollback
+			patchAndVerify(rbPatchStr, tc.containers, tc.expected, "rollback")
```

Comment:
It seems that we need more work to wait for a container whose `restartPolicy` is `RestartContainer` to be restarted at rollback because the container is restarted twice. The `RestartCount` in `tc.containers` should be `2` here. In addition, this code does not catch the second restart:

https://github.com/kubernetes/kubernetes/blob/6b73cccb07de30c7793370e1d5c3f8766d975d20/test/e2e/node/pod_resize.go#L442

I guess this causes a [flake](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/125202/pull-kubernetes-e2e-gce-cos-alpha-features/1799536236315021312):

```
Kubernetes e2e suite: [It] [sig-node] Pod InPlace Resize Container [Feature:InPlacePodVerticalScaling] Guaranteed QoS pod, one container - increase CPU (NotRequired) & memory (RestartContainer) expand_less 	18s
{ failed [FAILED] error running /home/prow/go/src/k8s.io/kubernetes/kubernetes/platforms/linux/amd64/kubectl --server=https://34.19.49.137 --kubeconfig=/workspace/.kube/config --namespace=pod-resize-9729 exec testpod --namespace=pod-resize-9729 --container=c1 -- head -n 1 /sys/fs/cgroup/memory/memory.limit_in_bytes:
Command stdout:

stderr:
head: /sys/fs/cgroup/memory/memory.limit_in_bytes: No such file or directory
command terminated with exit code 1

error:
exit status 1
In [It] at: k8s.io/kubernetes/test/e2e/framework/kubectl/builder.go:91 @ 06/08/24 21:22:23.903
}
```

This command could be called while the container is being restarted. Then, the command is failed and the cgoup v1 path is used:
https://github.com/kubernetes/kubernetes/blob/6b73cccb07de30c7793370e1d5c3f8766d975d20/test/e2e/node/pod_resize.go#L336

Reply:
Ah great catch @hshiina, will fix this!